chains/optimism/listener/listener.go

```diff
@@ -0,0 +1,133 @@
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ChainSafe/chainbridge-core/chains/evm/calls/evmclient"
+	"github.com/ChainSafe/chainbridge-core/types"
+
+	"github.com/ChainSafe/chainbridge-core/blockstore"
+	"github.com/ChainSafe/chainbridge-core/relayer/message"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/rs/zerolog/log"
+)
+
+var BlockRetryInterval = time.Second * 5
+
+// Only 1 for Optimism as block confirmations check occur when events are indexed in the data transport layer
+var BlockDelay = big.NewInt(5) //TODO: move to config
+
+type DepositLogs struct {
+	DestinationChainID uint8
+	ResourceID         [32]byte
+	DepositNonce       uint64
+}
+
+type ChainClient interface {
+	LatestBlock() (*big.Int, error)
+	FetchDepositLogs(ctx context.Context, address common.Address, startBlock *big.Int, endBlock *big.Int) ([]*evmclient.DepositLogs, error)
+	CallContract(ctx context.Context, callArgs map[string]interface{}, blockNumber *big.Int) ([]byte, error)
+	IsRollupVerified(blockNumber uint64) (bool, error)
+}
+
+type EventHandler interface {
+	HandleEvent(sourceID, destID uint8, nonce uint64, resourceID types.ResourceID, calldata, handlerResponse []byte) (*message.Message, error)
+}
+
+type EVMListener struct {
+	chainReader   ChainClient
+	eventHandler  EventHandler
+	bridgeAddress common.Address
+}
+
+func NewEVMListener(chainReader ChainClient, handler EventHandler, bridgeAddress common.Address) *EVMListener {
+	return &EVMListener{chainReader: chainReader, eventHandler: handler, bridgeAddress: bridgeAddress}
+}
+
+func (l *EVMListener) ListenToEvents(startBlock *big.Int, domainID uint8, kvrw blockstore.KeyValueWriter, stopChn <-chan struct{}, errChn chan<- error) <-chan *message.Message {
+	ch := make(chan *message.Message)
+	go func() {
+		for {
+			select {
+			case <-stopChn:
+				return
+			default:
+				head, err := l.chainReader.LatestBlock()
+				if err != nil {
+					log.Error().Err(err).Msg("Unable to get latest block")
+					time.Sleep(BlockRetryInterval)
+					continue
+				}
+
+				if startBlock == nil {
+					startBlock = head
+				}
+
+				// Sleep if the difference is less than BlockDelay; (latest - current) < BlockDelay
+				if big.NewInt(0).Sub(head, startBlock).Cmp(BlockDelay) == -1 {
+					time.Sleep(BlockRetryInterval)
+					continue
+				}
+
+				if verified, err := l.chainReader.IsRollupVerified(startBlock.Uint64()); err != nil {
+					log.Error().Err(err).Msg("Error while checking whether chain is verified")
+					log.Error().Msgf("Block Number: %v", startBlock)
+					time.Sleep(BlockRetryInterval)
+					continue
+				} else if !verified {
+					log.Error().Msg("Chain is not verified at current index")
+					log.Error().Msgf("Block Number: %v", startBlock)
+					time.Sleep(BlockRetryInterval)
+					continue
+				}
+				log.Info().Msgf("Verified index: %v", startBlock)
+
+				logs, err := l.chainReader.FetchDepositLogs(context.Background(), l.bridgeAddress, startBlock, startBlock)
```

Comment:
Would it be possible to check when fetching deposit logs is rollup verified and return rollup not verified error when it is not and then use standard EVM listener and chain modules and just replace EVM client with Optimism client to not copy logic?

Reply:
I had considered this when first writing the module. This should be possible and actually much cleaner thank you for bringing to my attention. I am going to make this change and test it. It has been a bit since I have altered the module since the EVM equivalence upgrade by Optimism so I do not remember exactly why I chose against this method, but I believe it had to do with how the core was formerly designed.

Reply:
Diving deeper I am starting to see some of why I had went away from this approach. The listener currently continues on an error from `FetchDepositLogs`. As per the comment in `listener.go` this is because `Filtering logs error really can appear only on wrong configuration or temporary network problem so i do no see any reason to break execution`, however we are now adding another reason for `FetchDepositLogs` to fail and would be continuing along the unverified chain and never sleeping to wait for a verified index. Because of this it looks like changing `FetchDepositLogs` for just the optimism client would still require having to change the listener to account for this fact.

Reply:
Before I saw your comment I had already begun changing the optimism client to return only the latest verified index for its `LatestBlock` call. In the current state of the listener, `LatestBlock` calls `eth_getBlockByNumber` which for optimism is the index of the last L2 tx processed not the last verified index. I think it might be cleaner to just have the optimism client's LatestBlock return the last verified index index. As you stated before with `FetchDepositLogs` this could be done within the constraints of the standard EVM listener and remove the need to copy logic with the added benefit that the optimistic listener will then only trace the verified optimism chain. 

What do you think?

Reply:
Yeah, looks much better now.